src/Utilities/FileSystem.cpp

```diff
@@ -0,0 +1,291 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Utilities/FileSystem.hpp"
+
+#include <cerrno>
+#include <chrono>
+#include <cmath>
+#include <cstdlib>
+#include <dirent.h>
+#include <libgen.h>
+#include <regex>
+#include <sstream>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <thread>
+#include <unistd.h>
+
+#include "ErrorHandling/Error.hpp"
+#include "Parallel/Printf.hpp"
+
+namespace filesystem {
+
+std::string dirname(const std::string& path) {
+  auto* file_path = new char[path.length() + 1];
+  strncpy(file_path, path.c_str(), path.length() + 1);
+  // The pointer from ::dirname is not deleted since it aliases with file_path
+  char* parent_dir_name = ::dirname(file_path);
+  std::string return_name(parent_dir_name);
+  delete[] file_path;
+  return return_name;
+}
+
+std::string get_file_name(const std::string& file_path) {
+  if (file_path.empty()) {
+    ERROR("Received an empty path");
+  }
+  if (file_path.find('/') == std::string::npos) {
+    // Handle file names such as 'dummy.txt' or '.dummy.txt'
+    return file_path;
+  }
+  std::smatch match{};
+  std::regex file_name_pattern{R"(^.*/(.+\..+))"};
```

Comment:
Not all files have a dot in their names, and some directories do, so this might return "foo.bar/baz".  How about `^.*/([^/]+)`?